File: golang/SubString/subarraySum/subarraySum.go
package subarraysum

func subarraysum(nums []int, k int) int {
	mp := map[int]int{}
	mp[0] = 1

	count := 0
	prefix := 0

	for _, num := range nums {
		prefix += num
		if _, ok := mp[prefix-k]; ok {
			count += mp[prefix-k]
		}

		mp[prefix]++
	}

	return count
}

File: cpp/SubString/subarraySum/subarraySum.cpp
#include <unordered_map>
#include <vector>

class Solution {
  public:
    int subarraySum(std::vector<int> nums, int k) {

        std::unordered_map<int, int> mp = {};
        mp[0] = 1;

        int count = 0;
        int prefix = 0;

        for (auto num : nums) {
            prefix += num;

            if (mp.find(prefix - k) != mp.end()) {
                count += mp[prefix - k];
            }

            mp[prefix]++;
        }

        return count;
    }
};
